File: packages/local/internal/projectindex/staticindex/protocol/completion.go
package protocol

// CompletionMethod identifies the cache-bypassing compiler query on the
// persistent Static Index worker.
const CompletionMethod = "completionQuery"

// CompletionPosition is a zero-based UTF-16 document position.
type CompletionPosition struct {
	Line      uint32 `json:"line"`
	Character uint32 `json:"character"`
}

// CompletionRange is a half-open UTF-16 document range.
type CompletionRange struct {
	Start CompletionPosition `json:"start"`
	End   CompletionPosition `json:"end"`
}

// CompletionCandidate is the compact Project Definition catalogue shape sent
// to the compiler. It intentionally contains no store or cache implementation
// detail.
type CompletionCandidate struct {
	ID          string `json:"id"`
	Kind        string `json:"kind"`
	Name        string `json:"name"`
	Binding     string `json:"binding"`
	File        string `json:"file"`
	Line        uint32 `json:"line,omitempty"`
	Character   uint32 `json:"character,omitempty"`
	Description string `json:"description,omitempty"`
}

// CompletionQuery is one unsaved source snapshot and its pinned candidates.
type CompletionQuery struct {
	File       string                `json:"file"`
	LanguageID string                `json:"languageId"`
	Source     string                `json:"source"`
	Position   CompletionPosition    `json:"position"`
	Candidates []CompletionCandidate `json:"candidates"`
	Limit      int                   `json:"limit"`
}

// CompletionWorkerRequest wraps one query in the persistent worker envelope.
type CompletionWorkerRequest struct {
	ID     uint64          `json:"id"`
	Method string          `json:"method"`
	Query  CompletionQuery `json:"query"`
}

// CompletionItem is one eager compiler-owned replacement recipe.
type CompletionItem struct {
	ID                  string               `json:"id"`
	Kind                string               `json:"kind"`
	Label               string               `json:"label"`
	Detail              string               `json:"detail"`
	InsertText          string               `json:"insertText"`
	Replacement         CompletionRange      `json:"replacement"`
	AdditionalTextEdits []CompletionTextEdit `json:"additionalTextEdits,omitempty"`
}

// CompletionTextEdit is one compiler-owned edit applied with the main
// replacement against the same unsaved document version.
type CompletionTextEdit struct {
	Range   CompletionRange `json:"range"`
	NewText string          `json:"newText"`
}

// CompletionResponse is the bounded transient compiler result.
type CompletionResponse struct {
	IsIncomplete bool             `json:"isIncomplete"`
	Items        []CompletionItem `json:"items"`
}

File: crates/protocol/src/completion.rs
//! Data-only request and response shapes for transient compiler completion.
//!
//! These values cross the persistent native-worker boundary. They are not
//! Project Index facts and must never be written to an index cache or patch.

use serde::{Deserialize, Serialize};

/// Method string for a transient completion query on the persistent worker.
pub const COMPLETION_QUERY_METHOD: &str = "completionQuery";

/// A zero-based UTF-16 source position.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionPosition {
    pub line: u32,
    pub character: u32,
}

/// A half-open UTF-16 source range.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionRange {
    pub start: CompletionPosition,
    pub end: CompletionPosition,
}

/// Compact Project Definition data admitted to a completion query.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionCandidate {
    pub id: String,
    pub kind: String,
    pub name: String,
    pub binding: String,
    pub file: String,
    #[serde(default)]
    pub line: u32,
    #[serde(default)]
    pub character: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// One additional eager source edit computed from the same unsaved parse.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionTextEdit {
    pub range: CompletionRange,
    pub new_text: String,
}

/// One cache-bypassing completion query over an unsaved document snapshot.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionQueryRequest {
    pub file: String,
    pub language_id: String,
    pub source: String,
    pub position: CompletionPosition,
    pub candidates: Vec<CompletionCandidate>,
    pub limit: usize,
}

/// One eager edit recipe produced by the compiler.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionQueryItem {
    pub id: String,
    pub kind: String,
    pub label: String,
    pub detail: String,
    pub insert_text: String,
    pub replacement: CompletionRange,
    #[serde(default)]
    pub additional_text_edits: Vec<CompletionTextEdit>,
}

/// Deterministic result from one transient compiler query.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionQueryResponse {
    pub is_incomplete: bool,
    pub items: Vec<CompletionQueryItem>,
}

/// Persistent-worker envelope for one completion query.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionWorkerRequest {
    pub id: u64,
    pub method: String,
    pub query: CompletionQueryRequest,
}
